### PR TITLE
refactor: condense v1 denom parsing to single function

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

closes: #6413 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **Refactor**
  - Simplified the extraction of denominations and traces in various functions by using a new, dedicated method.
  - Improved code clarity and modularity by delegating extraction logic to a separate function.

- **Bug Fixes**
  - Enhanced the validation process for prefixed denominations, ensuring more accurate and reliable handling of token denominations.

### CODE REVIEW

File: modules/apps/transfer/types/trace.go [hunk 1]
```diff
@@ -76,15 +67,23 @@ func (dt DenomTrace) GetFullDenomPath() string {
 	return dt.GetPrefix() + dt.BaseDenom
 }
 
-// extractPathAndBaseFromFullDenom returns the trace path and the base denom from
-// the elements that constitute the complete denom.
-func extractPathAndBaseFromFullDenom(fullDenomItems []string) ([]string, string) {
+// ExtractDenomFromFullPath returns the denom from the full path.
+// Used to support v1 denoms.
+func ExtractDenomFromFullPath(fullPath string) Denom {
```
Comment (colin-axner) [inline modules/apps/transfer/types/trace.go:72]: this should be the only function we need to go from: path -> Denom

File: modules/apps/transfer/types/trace.go [hunk 2]
```diff
@@ -95,17 +94,20 @@ func extractPathAndBaseFromFullDenom(fullDenomItems []string) ([]string, string)
 		// will be incorrectly parsed, but the token will continue to be treated correctly
 		// as an IBC denomination. The hash used to store the token internally on our chain
 		// will be the same value as the base denomination being correctly parsed.
-		if i < length-1 && length > 2 && channeltypes.IsValidChannelID(fullDenomItems[i+1]) {
-			pathSlice = append(pathSlice, fullDenomItems[i], fullDenomItems[i+1])
+		if i < length-1 && length > 2 && channeltypes.IsValidChannelID(denomSplit[i+1]) {
+			trace = append(trace, denomSplit[i]+"/"+denomSplit[i+1])
```
Comment (colin-axner) [inline modules/apps/transfer/types/trace.go:98]: will be typed soon enough

File: modules/apps/transfer/types/trace.go [hunk 3]
```diff
@@ -149,26 +151,20 @@ func (dt DenomTrace) Validate() error {
...
 	}
 
-	path := strings.Join(pathSlice, "/")
-	if path == "" {
-		// NOTE: base denom contains slashes, so no base denomination validation
-		return nil
+	path := strings.Join(denom.Trace, "/")
+	if len(denom.Trace) != 0 {
+		identifiers := strings.Split(path, "/")
+		return validateTraceIdentifiers(identifiers)
```
Comment (colin-axner) [inline modules/apps/transfer/types/trace.go:163]: soon `denom.Trace.Validate()`

File: modules/apps/transfer/types/trace.go (hunk 1 above)
Comment (crodriguezvega) [reply modules/apps/transfer/types/trace.go:72]: Maybe `FullPathToDenom` ([ref](https://github.com/cosmos/ibc-go/pull/6426#discussion_r1619414270)) or if we move this function to `denom.go` then maybe `FromFullPath` or `NewDenom`.

File: modules/apps/transfer/types/trace.go [hunk 4]
```diff
@@ -149,26 +151,20 @@ func (dt DenomTrace) Validate() error {
...
-func ValidatePrefixedDenom(denom string) error {
-	denomSplit := strings.Split(denom, "/")
-	if denomSplit[0] == denom && strings.TrimSpace(denom) != "" {
-		// NOTE: no base denomination validation
-		return nil
-	}
-
-	pathSlice, baseDenom := extractPathAndBaseFromFullDenom(denomSplit)
-	if strings.TrimSpace(baseDenom) == "" {
+func ValidatePrefixedDenom(fullPath string) error {
```
Comment (crodriguezvega) [inline modules/apps/transfer/types/trace.go:154]: I know this has already been like this for a while, but we also use `PrefixedDenom` and `FullPath` for trace + base denom. It would also be nice to converge now into a single term. I don't particularly like either, but what about `IBCDenom` (if we use on-chain denom for `ibc/<hash>`) or `ICS20Denom`?

**Review by crodriguezvega — APPROVED** [review]

File: modules/apps/transfer/types/trace.go (hunk 1 above)
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:72]: going to leave as is unless there's consensus otherwise

File: modules/apps/transfer/types/trace.go (hunk 4 above)
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:154]: I think it's good to open an issue and take discussion there. I'm happy to see the function renamed

**Review by chatton — CHANGES_REQUESTED** [review]
Everything looks good except the edge case I mentioned in the review, would be nice to make sure we have a test case covering this.

File: modules/apps/transfer/types/trace.go [hunk 5]
```diff
@@ -76,15 +67,23 @@ func (dt DenomTrace) GetFullDenomPath() string {
...
+// ExtractDenomFromFullPath returns the denom from the full path.
+// Used to support v1 denoms.
+func ExtractDenomFromFullPath(fullPath string) Denom {
+	denomSplit := strings.Split(fullPath, "/")
+
+	if denomSplit[0] == fullPath {
+		return Denom{
+			Base: fullPath,
+		}
+	}
```
[the remark this answers is not in the document]
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:79]: https://github.com/cosmos/ibc-go/blob/main/modules/apps/transfer/types/trace_test.go#L19

[the remark this answers is not in the document]
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:79]: This case is caught by the `baseDenomSlice` construction https://github.com/cosmos/ibc-go/pull/6424/files#diff-ff4125d8c59b9ab8785f0ae9e950dc2a59a79428482e3f36b7b1695f57d5e39fR105

[the remark this answers is not in the document]
Comment (chatton) [reply modules/apps/transfer/types/trace.go:79]: thinking now, it's possible that the correct behaviour ends up happening later in the function here

This actually ends up being handled correctly here `baseDenomSlice = denomSplit[i:]`

I feel like we should probably still explicitly check this at the start of the fn `if denomSplit[0] == fullPath` to be more explicit. I think this edge case could slip through if the for loop code gets changed in the future. We should at least add a test like

```golang
func TestFoo(t *testing.T) {
	res := types.ExtractDenomFromFullPath("atom/")
	require.Equal(t, "atom/", res.Base)
	require.Empty(t, res.Trace)
}
```

or similar WDYT?

[the remark this answers is not in the document]
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:79]: When we remove `ParseDenomTrace` we can move the test cases over to directly test the extraction function (I believe it didn't have unit tests since it was a private function and written before we made use of `export_test.go`

[the remark this answers is not in the document]
Comment (chatton) [reply modules/apps/transfer/types/trace.go:79]: ah yeah I hadn't thought of additional slashes inside, I guess it's not possible to do a check at the start, maybe we can add some additional tests for this function. Happy to do this in a follow up issue. It makes me a bit uneasy to have this happening in that for loop but I think there might be no alternative :thinking:

**Review by chatton — APPROVED** [review]
Leaving approval since the edge case I was concerned about is actually covered. We can add additional test coverage for this in a follow up :+1:

File: modules/apps/transfer/types/trace.go (hunk 4 above)
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:154]: this function will be deleted in a future pr
